src/Bedrock.Framework.Experimental/Protocols/Kafka/Messages/KafkaRequest.cs

```diff
@@ -0,0 +1,56 @@
+﻿#nullable enable
+
+using Bedrock.Framework.Experimental.Protocols.Kafka.Models;
+using Bedrock.Framework.Experimental.Protocols.Kafka.Primitives;
+using System;
+
+namespace Bedrock.Framework.Experimental.Protocols.Kafka.Messages.Requests
+{
+    public abstract class KafkaRequest : IDisposable
+    {
+        protected KafkaRequest(KafkaApiKeys apiKey, short apiVersion)
+        {
+            this.ApiKey = apiKey;
+            this.ApiVersion = apiVersion;
+        }
+
+        public KafkaApiKeys ApiKey { get; }
+        public short ApiVersion { get; }
+
+        public NullableString ClientId { get; set; }
+
+        public abstract void WriteRequest(ref PayloadWriter writer);
+
+        private bool disposedValue = false; // To detect redundant calls
+        protected virtual void Dispose(bool disposing)
+        {
+            if (!disposedValue)
+            {
+                if (disposing)
+                {
+                    // TODO: dispose managed state (managed objects).
+                }
+
+                // TODO: free unmanaged resources (unmanaged objects) and override a finalizer below.
+                // TODO: set large fields to null.
+
+                disposedValue = true;
+            }
+        }
+
+        // TODO: override a finalizer only if Dispose(bool disposing) above has code to free unmanaged resources.
```

Comment:
Personally I avoid writing finalizers in the vast majority of cases. It is the responsibility of the consumer to Dispose of disposables correctly. If they don't do that, they have a bug, and their code should be fixed.

See https://ericlippert.com/2015/05/18/when-everything-you-know-is-wrong-part-one/

Reply:
I didn't implement any Finalizers here, but I left the comments in that Visual Studio offers when implementing the dispose pattern for you.

Do you see anything wrong with the pattern here that was implemented?
I'll clean up the commented out finalizers that VS sets up...